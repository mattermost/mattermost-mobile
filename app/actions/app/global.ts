// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Tutorial} from '@constants';
import {GLOBAL_IDENTIFIERS} from '@constants/database';
import DatabaseManager from '@database/manager';
import {getActiveServerUrl} from '@init/credentials';
import {logError} from '@utils/log';

export const storeGlobal = async (id: string, value: unknown, prepareRecordsOnly = false) => {
    try {
        const {operator} = DatabaseManager.getAppDatabaseAndOperator();
        return operator.handleGlobal({
            globals: [{id, value}],
            prepareRecordsOnly,
        });
    } catch (error) {
        logError('storeGlobal', error);
        return {error};
    }
};

export const storeDeviceToken = async (token: string, prepareRecordsOnly = false) => {
    return storeGlobal(GLOBAL_IDENTIFIERS.DEVICE_TOKEN, token, prepareRecordsOnly);
};

export const storeOnboardingViewedValue = async (value = true) => {
    return storeGlobal(GLOBAL_IDENTIFIERS.ONBOARDING, value, false);
};

export const storeMultiServerTutorial = async (prepareRecordsOnly = false) => {
    return storeGlobal(Tutorial.MULTI_SERVER, 'true', prepareRecordsOnly);
};

export const storeProfileLongPressTutorial = async (prepareRecordsOnly = false) => {
    return storeGlobal(Tutorial.PROFILE_LONG_PRESS, 'true', prepareRecordsOnly);
};

export const storeSkinEmojiSelectorTutorial = async (prepareRecordsOnly = false) => {
    return storeGlobal(Tutorial.EMOJI_SKIN_SELECTOR, 'true', prepareRecordsOnly);
};

export const storeDraftsTutorial = async () => {
    return storeGlobal(Tutorial.DRAFTS, 'true', false);
};

export const storeDontAskForReview = async (prepareRecordsOnly = false) => {
    return storeGlobal(GLOBAL_IDENTIFIERS.DONT_ASK_FOR_REVIEW, 'true', prepareRecordsOnly);
};

export const storeLastAskForReview = async (prepareRecordsOnly = false) => {
    return storeGlobal(GLOBAL_IDENTIFIERS.LAST_ASK_FOR_REVIEW, Date.now(), prepareRecordsOnly);
};

export const storeFirstLaunch = async (prepareRecordsOnly = false) => {
    return storeGlobal(GLOBAL_IDENTIFIERS.FIRST_LAUNCH, Date.now(), prepareRecordsOnly);
};

export const storeLastViewedChannelIdAndServer = async (channelId: string) => {
    const currentServerUrl = await getActiveServerUrl();

    return storeGlobal(GLOBAL_IDENTIFIERS.LAST_VIEWED_CHANNEL, {
        server_url: currentServerUrl,
        channel_id: channelId,
    }, false);
};

export const storeLastViewedThreadIdAndServer = async (threadId: string) => {
    const currentServerUrl = await getActiveServerUrl();

    return storeGlobal(GLOBAL_IDENTIFIERS.LAST_VIEWED_THREAD, {
        server_url: currentServerUrl,
        thread_id: threadId,
    }, false);
};

export const removeLastViewedChannelIdAndServer = async () => {
    return storeGlobal(GLOBAL_IDENTIFIERS.LAST_VIEWED_CHANNEL, null, false);
};

export const removeLastViewedThreadIdAndServer = async () => {
    return storeGlobal(GLOBAL_IDENTIFIERS.LAST_VIEWED_THREAD, null, false);
};

export const storePushDisabledInServerAcknowledged = async (serverUrl: string) => {
    return storeGlobal(`${GLOBAL_IDENTIFIERS.PUSH_DISABLED_ACK}${serverUrl}`, 'true', false);
};

export const removePushDisabledInServerAcknowledged = async (serverUrl: string) => {
    return storeGlobal(`${GLOBAL_IDENTIFIERS.PUSH_DISABLED_ACK}${serverUrl}`, null, false);
};
