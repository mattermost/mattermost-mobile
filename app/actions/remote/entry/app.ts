// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {setLastServerVersionCheck} from '@actions/local/systems';
import {fetchConfigAndLicense} from '@actions/remote/systems';
import DatabaseManager from '@database/manager';
import WebsocketManager from '@managers/websocket_manager';
import {prepareCommonSystemValues} from '@queries/servers/system';
import {deleteV1Data} from '@utils/file';

import {verifyPushProxy} from './common';

export async function appEntry(serverUrl: string, since = 0) {
    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
    if (!operator) {
        return {error: `${serverUrl} database not found`};
    }

    if (!since) {
        if (Object.keys(DatabaseManager.serverDatabases).length === 1) {
            await setLastServerVersionCheck(serverUrl, true);
        }
    }

    // clear lastUnreadChannelId
    const removeLastUnreadChannelId = await prepareCommonSystemValues(operator, {lastUnreadChannelId: ''});
    if (removeLastUnreadChannelId) {
        await operator.batchRecords(removeLastUnreadChannelId, 'appEntry - removeLastUnreadChannelId');
    }

    WebsocketManager.openAll('entry');

    verifyPushProxy(serverUrl);

    return {};
}

export async function upgradeEntry(serverUrl: string) {
    const dt = Date.now();

    try {
        const configAndLicense = await fetchConfigAndLicense(serverUrl, false);
        const entryData = await appEntry(serverUrl, 0);
        const error = configAndLicense.error || entryData.error;

        if (!error) {
            await DatabaseManager.updateServerIdentifier(serverUrl, configAndLicense.config!.DiagnosticId);
            await DatabaseManager.setActiveServerDatabase(serverUrl);
            deleteV1Data();
        }

        return {error, time: Date.now() - dt};
    } catch (e) {
        return {error: e, time: Date.now() - dt};
    }
}
