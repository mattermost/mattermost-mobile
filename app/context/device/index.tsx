// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import RNUtils, {type SplitViewResult} from '@mattermost/rnutils';
import React, {createContext, useEffect, useState} from 'react';
import {NativeEventEmitter} from 'react-native';

type Props = {
    children: React.ReactNode;
}

const emitter = new NativeEventEmitter(RNUtils);

let info = RNUtils.isRunningInSplitView();

export const DeviceContext = createContext(info);
const {Provider} = DeviceContext;

const DeviceInfoProvider = ({children}: Props) => {
    const [deviceInfo, setDeviceInfo] = useState(info);
    useEffect(() => {
        const listener = emitter.addListener('SplitViewChanged', (result: SplitViewResult) => {
            setDeviceInfo(result);
            info = result;
        });

        return () => listener.remove();
    }, []);

    return (
        <Provider value={deviceInfo}>
            {children}
        </Provider>);
};

export default DeviceInfoProvider;
