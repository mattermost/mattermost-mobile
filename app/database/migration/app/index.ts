// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {schemaMigrations} from '@nozbe/watermelondb/Schema/migrations';

// NOTE : To implement migration, please follow this document
// https://nozbe.github.io/WatermelonDB/Advanced/Migrations.html

export default schemaMigrations({migrations: []});
