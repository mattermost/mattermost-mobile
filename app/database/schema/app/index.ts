// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {type AppSchema, appSchema, tableSchema} from '@nozbe/watermelondb';

import {MM_TABLES} from '@constants/database';

const {INFO, GLOBAL, SERVERS} = MM_TABLES.APP;

export const schema: AppSchema = appSchema({
    version: 1,
    tables: [
        tableSchema({
            name: INFO,
            columns: [
                {name: 'build_number', type: 'string'},
                {name: 'created_at', type: 'number'},
                {name: 'version_number', type: 'string'},
            ],
        }),
        tableSchema({
            name: GLOBAL,
            columns: [
                {name: 'value', type: 'string'},
            ],
        }),
        tableSchema({
            name: SERVERS,
            columns: [
                {name: 'db_path', type: 'string'},
                {name: 'display_name', type: 'string'},
                {name: 'identifier', type: 'string', isIndexed: true},
                {name: 'last_active_at', type: 'number', isIndexed: true},
                {name: 'url', type: 'string', isIndexed: true},
            ],
        }),
    ],
});
