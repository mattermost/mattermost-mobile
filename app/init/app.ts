// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {CallsManager} from '@calls/calls_manager';
import DatabaseManager from '@database/manager';
import {getAllServerCredentials} from '@init/credentials';
import {initialLaunch} from '@init/launch';
import ManagedApp from '@init/managed_app';
import PushNotifications from '@init/push_notifications';
import GlobalEventHandler from '@managers/global_event_handler';
import NetworkManager from '@managers/network_manager';
import SessionManager from '@managers/session_manager';
import WebsocketManager from '@managers/websocket_manager';
import {registerScreens} from '@screens/index';
import {registerNavigationListeners} from '@screens/navigation';
import EphemeralStore from '@store/ephemeral_store';
import NavigationStore from '@store/navigation_store';

// Controls whether the main initialization (database, etc...) is done, either on app launch
// or on the Share Extension, for example.
let baseAppInitialized = false;

let serverCredentials: ServerCredential[];

// Fallback Polyfill for Promise.allSettle
Promise.allSettled = Promise.allSettled || (<T>(promises: Array<Promise<T>>) => Promise.all(
    promises.map((p) => p.
        then((value) => ({
            status: 'fulfilled',
            value,
        })).
        catch((reason) => ({
            status: 'rejected',
            reason,
        })),
    ),
));

export async function initialize() {
    if (!baseAppInitialized) {
        baseAppInitialized = true;
        serverCredentials = await getAllServerCredentials();
        const serverUrls = serverCredentials.map((credential) => credential.serverUrl);

        await DatabaseManager.init(serverUrls);
        await NetworkManager.init(serverCredentials);

        GlobalEventHandler.init();
        ManagedApp.init();
        SessionManager.init();
        CallsManager.initialize();
    }
}

export async function start() {
    // Clean relevant information on ephemeral stores
    NavigationStore.reset();
    EphemeralStore.setCurrentThreadId('');
    EphemeralStore.setProcessingNotification('');

    await initialize();

    PushNotifications.init(serverCredentials.length > 0);

    registerNavigationListeners();
    registerScreens();

    await WebsocketManager.init(serverCredentials);

    initialLaunch();
}
