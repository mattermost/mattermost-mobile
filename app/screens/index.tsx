// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Provider as EMMProvider} from '@mattermost/react-native-emm';
import React, {type ComponentType} from 'react';
import {IntlProvider} from 'react-intl';
import {Platform, type StyleProp, type ViewStyle} from 'react-native';
import {GestureHandlerRootView} from 'react-native-gesture-handler';
import {Navigation} from 'react-native-navigation';
import {SafeAreaProvider} from 'react-native-safe-area-context';

import {Screens} from '@constants';
import {withServerDatabase} from '@database/components';
import {DEFAULT_LOCALE, getTranslations} from '@i18n';

const withGestures = (Screen: React.ComponentType, styles: StyleProp<ViewStyle>) => {
    return function gestureHoc(props: any) {
        return (
            <GestureHandlerRootView style={[{flex: 1}, styles]}>
                <Screen {...props}/>
            </GestureHandlerRootView>
        );
    };
};

const withIntl = (Screen: React.ComponentType) => {
    return function IntlEnabledComponent(props: any) {
        return (
            <IntlProvider
                locale={DEFAULT_LOCALE}
                messages={getTranslations(DEFAULT_LOCALE)}
            >
                <Screen {...props}/>
            </IntlProvider>
        );
    };
};

const withSafeAreaInsets = (Screen: React.ComponentType) => {
    return function SafeAreaInsets(props: any) {
        return (
            <SafeAreaProvider>
                <Screen {...props}/>
            </SafeAreaProvider>
        );
    };
};

const withManagedConfig = (Screen: React.ComponentType) => {
    return function EmmProvider(props: any) {
        return (
            <EMMProvider>
                <Screen {...props}/>
            </EMMProvider>
        );
    };
};

Navigation.setLazyComponentRegistrator((screenName) => {
    let screen: any|undefined;
    let extraStyles: StyleProp<ViewStyle>;
    switch (screenName) {
        case Screens.ABOUT:
            screen = withServerDatabase(require('@screens/settings/about').default);
            break;
        case Screens.APPS_FORM:
            screen = withServerDatabase(require('@screens/apps_form').default);
            break;
        case Screens.BOTTOM_SHEET:
            screen = withServerDatabase(require('@screens/bottom_sheet').default);
            Navigation.registerComponent(Screens.BOTTOM_SHEET, () =>
                withGestures(withSafeAreaInsets(withManagedConfig(screen)), undefined),
            );
            return;
        case Screens.BROWSE_CHANNELS:
            screen = withServerDatabase(require('@screens/browse_channels').default);
            break;
        case Screens.CHANNEL:
            screen = withServerDatabase(require('@screens/channel').default);
            break;
        case Screens.CHANNEL_BOOKMARK:
            screen = withServerDatabase(require('@screens/channel_bookmark').default);
            break;
        case Screens.CHANNEL_NOTIFICATION_PREFERENCES:
            screen = withServerDatabase(require('@screens/channel_notification_preferences').default);
            break;
        case Screens.CHANNEL_FILES:
            screen = withServerDatabase(require('@screens/channel_files').default);
            break;
        case Screens.CHANNEL_INFO:
            screen = withServerDatabase(require('@screens/channel_info').default);
            break;
        case Screens.CODE:
            screen = withServerDatabase(require('@screens/code').default);
            break;
        case Screens.CONVERT_GM_TO_CHANNEL:
            screen = withServerDatabase(require('@screens/convert_gm_to_channel').default);
            break;
        case Screens.CREATE_OR_EDIT_CHANNEL:
            screen = withServerDatabase(require('@screens/create_or_edit_channel').default);
            break;
        case Screens.CUSTOM_STATUS:
            screen = withServerDatabase(require('@screens/custom_status').default);
            break;
        case Screens.CUSTOM_STATUS_CLEAR_AFTER:
            screen = withServerDatabase(require('@screens/custom_status_clear_after').default);
            break;
        case Screens.CREATE_DIRECT_MESSAGE:
            screen = withServerDatabase(require('@screens/create_direct_message').default);
            break;
        case Screens.CHANNEL_ADD_MEMBERS:
            screen = withServerDatabase(require('@screens/channel_add_members').default);
            break;
        case Screens.EDIT_POST:
            screen = withServerDatabase(require('@screens/edit_post').default);
            break;
        case Screens.EDIT_PROFILE:
            screen = withServerDatabase(require('@screens/edit_profile').default);
            break;
        case Screens.EDIT_SERVER:
            screen = withIntl(require('@screens/edit_server').default);
            break;
        case Screens.EMOJI_PICKER:
            screen = withServerDatabase(require('@screens/emoji_picker').default);
            break;
        case Screens.FIND_CHANNELS:
            screen = withServerDatabase(require('@screens/find_channels').default);
            break;
        case Screens.FORGOT_PASSWORD:
            screen = withIntl(require('@screens/forgot_password').default);
            break;
        case Screens.GALLERY:
            screen = withServerDatabase(require('@screens/gallery').default);
            break;
        case Screens.GENERIC_OVERLAY:
            screen = withServerDatabase(require('@screens/overlay').default);
            break;
        case Screens.GLOBAL_THREADS:
            screen = withServerDatabase(require('@screens/global_threads').default);
            break;
        case Screens.INTERACTIVE_DIALOG:
            screen = withServerDatabase(require('@screens/interactive_dialog').default);
            break;
        case Screens.INTEGRATION_SELECTOR:
            screen = withServerDatabase(require('@screens/integration_selector').default);
            break;
        case Screens.INVITE:
            screen = withServerDatabase(require('@screens/invite').default);
            break;
        case Screens.IN_APP_NOTIFICATION: {
            const notificationScreen = require('@screens/in_app_notification').default;
            Navigation.registerComponent(Screens.IN_APP_NOTIFICATION, () =>
                Platform.select({
                    default: notificationScreen,
                    ios: withSafeAreaInsets(notificationScreen),
                }),
            );
            return;
        }
        case Screens.JOIN_TEAM:
            screen = withServerDatabase(require('@screens/join_team').default);
            break;
        case Screens.LATEX:
            screen = withServerDatabase(require('@screens/latex').default);
            break;
        case Screens.LOGIN:
            screen = withIntl(require('@screens/login').default);
            break;
        case Screens.MANAGE_CHANNEL_MEMBERS:
            screen = withServerDatabase(require('@screens/manage_channel_members').default);
            break;
        case Screens.MFA:
            screen = withIntl(require('@screens/mfa').default);
            break;
        case Screens.SELECT_TEAM:
            screen = withServerDatabase(require('@screens/select_team').default);
            break;
        case Screens.PERMALINK:
            screen = withServerDatabase(require('@screens/permalink').default);
            break;
        case Screens.PINNED_MESSAGES:
            screen = withServerDatabase(require('@screens/pinned_messages').default);
            break;
        case Screens.POST_OPTIONS:
            screen = withServerDatabase(require('@screens/post_options').default);
            break;
        case Screens.POST_PRIORITY_PICKER:
            screen = withServerDatabase(require('@screens/post_priority_picker').default);
            break;
        case Screens.REACTIONS:
            screen = withServerDatabase(require('@screens/reactions').default);
            break;
        case Screens.REVIEW_APP:
            screen = withServerDatabase(require('@screens/review_app').default);
            break;
        case Screens.SETTINGS:
            screen = withServerDatabase(require('@screens/settings').default);
            break;
        case Screens.SETTINGS_ADVANCED:
            screen = withServerDatabase(require('@screens/settings/advanced').default);
            break;
        case Screens.SETTINGS_DISPLAY:
            screen = withServerDatabase(require('@screens/settings/display').default);
            break;
        case Screens.SETTINGS_DISPLAY_CLOCK:
            screen = withServerDatabase(require('@screens/settings/display_clock').default);
            break;
        case Screens.SETTINGS_DISPLAY_CRT:
            screen = withServerDatabase(require('@screens/settings/display_crt').default);
            break;
        case Screens.SETTINGS_DISPLAY_THEME:
            screen = withServerDatabase(require('@screens/settings/display_theme').default);
            break;
        case Screens.SETTINGS_DISPLAY_TIMEZONE:
            screen = withServerDatabase(require('@screens/settings/display_timezone').default);
            break;
        case Screens.SETTINGS_DISPLAY_TIMEZONE_SELECT:
            screen = withServerDatabase(require('@screens/settings/display_timezone_select').default);
            break;
        case Screens.SETTINGS_NOTIFICATION:
            screen = withServerDatabase(require('@screens/settings/notifications').default);
            break;
        case Screens.SETTINGS_NOTIFICATION_AUTO_RESPONDER:
            screen = withServerDatabase(require('@screens/settings/notification_auto_responder').default);
            break;
        case Screens.SETTINGS_NOTIFICATION_EMAIL:
            screen = withServerDatabase(require('@screens/settings/notification_email').default);
            break;
        case Screens.SETTINGS_NOTIFICATION_MENTION:
            screen = withServerDatabase(require('@screens/settings/notification_mention').default);
            break;
        case Screens.SETTINGS_NOTIFICATION_PUSH:
            screen = withServerDatabase(require('@screens/settings/notification_push').default);
            break;
        case Screens.SETTINGS_NOTIFICATION_CALL:
            screen = withServerDatabase(require('@screens/settings/notification_call').default);
            break;
        case Screens.SHARE_FEEDBACK:
            screen = withServerDatabase(require('@screens/share_feedback').default);
            break;
        case Screens.SNACK_BAR: {
            const snackBarScreen = withServerDatabase(require('@screens/snack_bar').default);
            Navigation.registerComponent(Screens.SNACK_BAR, () =>
                Platform.select({
                    default: snackBarScreen,
                    ios: withSafeAreaInsets(snackBarScreen) as ComponentType,
                }),
            );
            break;
        }
        case Screens.SSO:
            screen = withIntl(require('@screens/sso').default);
            break;
        case Screens.TABLE:
            screen = withServerDatabase(require('@screens/table').default);
            break;
        case Screens.TEAM_SELECTOR_LIST:
            screen = withServerDatabase(require('@screens/convert_gm_to_channel/team_selector_list').default);
            break;
        case Screens.TERMS_OF_SERVICE:
            screen = withServerDatabase(require('@screens/terms_of_service').default);
            break;
        case Screens.THREAD:
            screen = withServerDatabase(require('@screens/thread').default);
            break;
        case Screens.THREAD_FOLLOW_BUTTON:
            Navigation.registerComponent(Screens.THREAD_FOLLOW_BUTTON, () => withServerDatabase(
                require('@screens/thread/thread_follow_button').default,
            ));
            break;
        case Screens.THREAD_OPTIONS:
            screen = withServerDatabase(require('@screens/thread_options').default);
            break;
        case Screens.USER_PROFILE:
            screen = withServerDatabase(require('@screens/user_profile').default);
            break;
        case Screens.CALL:
            screen = withServerDatabase(require('@calls/screens/call_screen').default);
            break;
        case Screens.CALL_PARTICIPANTS:
            screen = withServerDatabase(require('@calls/screens/participants_list').default);
            break;
        case Screens.CALL_HOST_CONTROLS:
            screen = withServerDatabase(require('@calls/screens/host_controls').default);
            break;
    }

    if (screen) {
        Navigation.registerComponent(screenName, () => withGestures(withSafeAreaInsets(withManagedConfig(screen)), extraStyles));
    }
});

export function registerScreens() {
    const homeScreen = require('@screens/home').default;
    const serverScreen = require('@screens/server').default;
    const onboardingScreen = require('@screens/onboarding').default;
    Navigation.registerComponent(Screens.ONBOARDING, () => withGestures(withIntl(withManagedConfig(onboardingScreen)), undefined));
    Navigation.registerComponent(Screens.SERVER, () => withSafeAreaInsets(withGestures(withIntl(withManagedConfig(serverScreen)), undefined)));
    Navigation.registerComponent(Screens.HOME, () => withGestures(withSafeAreaInsets(withServerDatabase(withManagedConfig(homeScreen))), undefined));
}
