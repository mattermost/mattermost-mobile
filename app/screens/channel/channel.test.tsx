// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {act, screen} from '@testing-library/react-native';
import {Navigation} from 'react-native-navigation';

import {storeLastViewedChannelIdAndServer} from '@actions/app/global';
import {useChannelSwitch} from '@hooks/channel_switch';
import {useIsTablet} from '@hooks/device';
import {useTeamSwitch} from '@hooks/team_switch';
import {renderWithEverything} from '@test/intl-test-helper';
import TestHelper from '@test/test_helper';

import Channel from './channel';

import type Database from '@nozbe/watermelondb/Database';
import type {AvailableScreens} from '@typings/screens/navigation';

jest.mock('@actions/app/global', () => ({
    storeLastViewedChannelIdAndServer: jest.fn(),
    removeLastViewedChannelIdAndServer: jest.fn(),
}));

jest.mock('@hooks/device', () => ({
    useIsTablet: jest.fn().mockReturnValue(false),
}));

jest.mock('@hooks/channel_switch', () => ({
    useChannelSwitch: jest.fn().mockReturnValue(false),
}));

jest.mock('@hooks/team_switch', () => ({
    useTeamSwitch: jest.fn().mockReturnValue(false),
}));

jest.mock('@react-native-camera-roll/camera-roll', () => ({
    CameraRoll: {
        save: jest.fn().mockResolvedValue('path'),
    },
}));

jest.mock('react-native-navigation', () => ({
    Navigation: {
        events: jest.fn().mockReturnValue({
            registerComponentListener: jest.fn().mockReturnValue({
                remove: jest.fn(),
            }),
        }),
    },
}));

function getBaseProps() {
    return {
        channelId: 'channel-id',
        componentId: 'component-id' as AvailableScreens,
        showJoinCallBanner: false,
        isInACall: false,
        isCallsEnabledInChannel: true,
        groupCallsAllowed: true,
        showIncomingCalls: false,
        isTabletView: false,
        dismissedGMasDMNotice: [],
        currentUserId: 'current-user-id',
        channelType: 'O' as ChannelType,
        hasGMasDMFeature: true,
        includeBookmarkBar: false,
    };
}

describe('Channel', () => {
    let database: Database;

    beforeAll(async () => {
        const server = await TestHelper.setupServerDatabase();
        database = server.database;
    });

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should match snapshot', async () => {
        const wrapper = renderWithEverything(
            <Channel {...getBaseProps()}/>,
            {database},
        );
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('renders channel screen correctly', async () => {
        let wrapper;
        await act(async () => {
            wrapper = renderWithEverything(
                <Channel {...getBaseProps()}/>,
                {database},
            );
        });

        expect(wrapper.getByTestId('channel.screen')).toBeTruthy();
        expect(storeLastViewedChannelIdAndServer).toHaveBeenCalledWith(getBaseProps().channelId);
        expect(Navigation.events().registerComponentListener).toHaveBeenCalled();
    });

    it('shows floating call container when in a call', async () => {
        const props = getBaseProps();
        props.isInACall = true;

        renderWithEverything(
            <Channel {...props}/>,
            {database},
        );

        // Wait for posts to render
        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 0));
        });

        expect(screen.getByTestId('floating_call_container')).toBeTruthy();
    });

    it('shows floating call container with join banner', async () => {
        const props = getBaseProps();
        props.showJoinCallBanner = true;

        renderWithEverything(
            <Channel {...props}/>,
            {database},
        );

        // Wait for posts to render
        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 0));
        });

        expect(screen.getByTestId('floating_call_container')).toBeTruthy();
    });

    it('handles tablet view correctly', async () => {
        const mockedIsTablet = jest.mocked(useIsTablet);
        mockedIsTablet.mockReturnValue(true);

        const props = getBaseProps();
        props.isTabletView = true;

        renderWithEverything(
            <Channel {...props}/>,
            {database},
        );

        expect(screen.getByTestId('channel.screen')).toBeTruthy();
    });

    it('does not render posts while switching teams', async () => {
        const mockedTeamSwitch = jest.mocked(useTeamSwitch);
        mockedTeamSwitch.mockReturnValue(true);

        renderWithEverything(
            <Channel {...getBaseProps()}/>,
            {database},
        );

        expect(screen.queryByTestId('channel.post_draft')).toBeNull();
    });

    it('does not render posts while switching channels', async () => {
        const mockedChannelSwitch = jest.mocked(useChannelSwitch);
        mockedChannelSwitch.mockReturnValue(true);

        renderWithEverything(
            <Channel {...getBaseProps()}/>,
            {database},
        );

        expect(screen.queryByTestId('channel.post_draft')).toBeNull();
    });
});
