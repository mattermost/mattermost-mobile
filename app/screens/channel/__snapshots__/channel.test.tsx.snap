// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Channel should match snapshot 1`] = `
<View
  style={
    {
      "height": "100%",
      "position": "absolute",
      "width": "100%",
    }
  }
>
  <RNCSafeAreaView
    edges={
      {
        "bottom": "off",
        "left": "additive",
        "right": "additive",
        "top": "off",
      }
    }
    mode="margin"
    onLayout={[Function]}
    style={
      {
        "flex": 1,
      }
    }
    testID="channel.screen"
  >
    <View
      style={
        [
          {
            "backgroundColor": "#1e325c",
            "position": "absolute",
            "width": "100%",
            "zIndex": 10,
          },
          {
            "height": 44,
            "maxHeight": 208,
            "minHeight": 44,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "#1e325c",
              "flexDirection": "row",
              "justifyContent": "flex-start",
              "paddingHorizontal": 16,
              "zIndex": 10,
            },
            {
              "height": 44,
              "paddingTop": 0,
            },
          ]
        }
      >
        <View
          style={
            {
              "bottom": 0,
              "height": "100%",
              "justifyContent": "center",
              "paddingLeft": 16,
              "position": "absolute",
              "zIndex": 5,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            hitSlop={
              {
                "bottom": 20,
                "left": 20,
                "right": 20,
                "top": 20,
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="navigation.header.back"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Icon
                color="#ffffff"
                name="arrow-back-ios"
                size={24}
              />
              <View
                style={
                  {
                    "minWidth": 24,
                  }
                }
              />
            </View>
          </View>
        </View>
        <View
          style={
            [
              {
                "alignItems": "center",
                "bottom": 0,
                "flex": undefined,
                "height": "100%",
                "justifyContent": "center",
                "left": 16,
                "paddingHorizontal": 60,
                "position": "absolute",
                "width": "100%",
                "zIndex": 1,
              },
              {
                "marginLeft": undefined,
              },
            ]
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                }
              }
            >
              <Text
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  [
                    {
                      "color": "#ffffff",
                      "fontFamily": "Metropolis-SemiBold",
                      "fontSize": 18,
                      "fontWeight": "600",
                      "lineHeight": 24,
                    },
                    {
                      "opacity": 1,
                    },
                  ]
                }
                testID="navigation.header.title"
              />
              <View
                style={
                  {
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "left": undefined,
                  }
                }
              >
                <Text
                  ellipsizeMode="tail"
                  numberOfLines={1}
                  style={
                    {
                      "color": "rgba(255,255,255,0.72)",
                      "fontFamily": "OpenSans",
                      "fontSize": 12,
                      "fontWeight": "400",
                      "height": 13,
                      "lineHeight": 12,
                      "marginBottom": 8,
                      "marginTop": 2,
                    }
                  }
                  testID="navigation.header.subtitle"
                >
                  View info
                </Text>
                <Icon
                  color="rgba(255,255,255,0.72)"
                  name="chevron-right"
                  size={14}
                />
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "bottom": 0,
              "flexDirection": "row",
              "height": "100%",
              "justifyContent": "flex-end",
              "position": "absolute",
              "right": 16,
              "zIndex": 2,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            hitSlop={
              {
                "bottom": 5,
                "left": 5,
                "right": 5,
                "top": 20,
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="channel_header.channel_quick_actions.button"
          >
            <Icon
              color="#ffffff"
              name="dots-horizontal"
              size={24}
            />
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "top": 44,
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "#1e325c",
            "height": 40,
            "position": "absolute",
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "#ffffff",
              "borderTopLeftRadius": 12,
              "borderTopRightRadius": 12,
              "flex": 1,
            }
          }
        />
      </View>
    </View>
  </RNCSafeAreaView>
</View>
`;
